22008-beta draft (#2367)
OpenLiberty/blogs: posts/2022-07-07-beta-release-post.adoc

open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Enhance the effectiveness of log analysis tools with new JSON fields for logged exceptions in Open Liberty 22.0.0.8-beta
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

Open Liberty 22.0.0.8 offers new JSON fields for logged exceptions, enhancing the Open Liberty JSON logging and logstashCollector-1.0 features, by increasing the effectiveness of log analysis tools down stream, as rules and visualizations relevant to these two new exception related fields can be created which results in more stream-lined strategy to identify any issues encountered by the application.

The link:{url-about}[Open Liberty] 22.0.0.8-beta includes the following beta features (along with all GA features):

* <<json_fields, New JSON fields for logged exceptions (via parameters)>>

[#json_fields]
== New JSON fields for logged exceptions (via parameters)

> Incorporated as part of the Java's logger API are methods that allow you to include a `Throwable` as a parameter. This new functionality introduced enhances the Open Liberty's JSON logging and the `logstashCollector-1.0` feature (which forwards JSON logs to an Elasticsearch target server) by introducing two new fields relevant to the `Throwable` object. For OpenLiberty's JSON logging the fields are `ibm_stackTrace` and `ibm_exceptionName`. For the `logstashCollector-1.0` feature the fields are `stackTrace` and `exceptionName`. The _stack trace_ field would present solely the stack trace of the `Throwable`. The _exception name_ field would present the type of exception of the `Throwable`. These fields are only included if the logging record includes a `Throwable` object. This affects the message and trace events for both the OpenLiberty JSON logging and `logstashCollector-1.0` features.

Reviewer: ```suggestion
Incorporated as part of the Java's logger API are methods that allow you to include a `Throwable` as a parameter. This new functionality enhances Open Liberty's JSON logging and the link:{url-prefix}/docs/latest/reference/feature/logstashCollector-1.0.html[`logstashCollector-1.0` feature] by introducing two new fields relevant to the `Throwable` object. For Open Liberty's JSON logging, the fields are `ibm_stackTrace` and `ibm_exceptionName`. For the `logstashCollector-1.0` feature the fields are `stackTrace` and `exceptionName`. The `stack trace` field would present solely the stack trace of the `Throwable`. The `exception name` field would present the type of exception of the `Throwable`. These fields are only included if the logging record includes a `Throwable` object. This affects the message and trace events for both the Open Liberty JSON logging and `logstashCollector-1.0` features.
```

> Prior to this enhanced functionality, any logging record originating from a Java Logger which made use of any of the methods that accept a `Throwable` parameter would simply append the stack trace to the existing `message` field.  By keeping the `message` field solely to the logged message and having a unique field for the stack trace and exception type enhances the effectiveness of log analysis tools down stream. Rules and visualizations relevant to these two new exception related fields can be created which results in more stream-lined strategy to identify any issues encountered by the application. There are no configuration steps to enable this functionality and will be present in the OpenLiberty's JSON logging and `logstashCollector-1.0` features.

Reviewer: ```suggestion
Prior to this enhancement, any logging record originating from a Java Logger which made use of any of the methods that accept a `Throwable` parameter would simply append the stack trace to the existing `message` field.  By keeping the `message` field solely for the logged message and having a unique field for the stack trace and exception type enhances the effectiveness of log analysis tools down stream. Rules and visualizations relevant to these two new exception related fields can be created which results in more stream-lined strategy to identify any issues encountered by the application. No configuration is necessary to enable the new functionality.
```

*Example:*

> For the example we will only be showing the message log record for OpenLiberty's JSON logging.

Reviewer: ```suggestion
For the example we will only be showing the message log record for Open Liberty's JSON logging.
```

Application with the following code snippet:
[source, java]
----
Logger logger = Logger.getLogger(MyResource.class.getCanonicalName());
Exception exception = new IllegalArgumentException("ouch");
logger.log(Level.INFO, "exception message", exception);
----

> This is the OpenLiberty JSON logging output before:

Reviewer: ```suggestion
Previous Open Liberty JSON logging output:
```

[source, json]
----
{
    "type": "liberty_message",
    ...
    "message": “exception message  java.lang.RuntimeException: ouch\r\n\tatmy.package.MyResource.get(MyResource.java:32)\r\n\tatmy.package.MyResource.get(MyResource.java:20)\r\n...",
    ...
}
----

> This is the OpenLiberty JSON logging output after:

Reviewer: ```suggestion
New Open Liberty JSON logging output:
```
